stl/inc/ranges

```diff
@@ -22,15 +25,461 @@ _STL_DISABLE_CLANG_WARNINGS
+        }
+
+        constexpr _Ty& operator*() noexcept {
+            return _Val;
+        }
+        constexpr const _Ty& operator*() const noexcept {
+            return _Val;
+        }
+
+    private:
+        /* [[no_unique_address]] */ _Ty _Val = _Ty();
+    };
+#endif // TRANSITION, VSO-1174090
+
+    // CLASS TEMPLATE ranges::empty_view
+    // clang-format off
+    template <class _Ty>
+        requires is_object_v<_Ty>
+    class empty_view : public view_interface<empty_view<_Ty>> {
+        // clang-format on
+    public:
+        static constexpr _Ty* begin() noexcept {
+            return nullptr;
+        }
+
+        static constexpr _Ty* end() noexcept {
+            return nullptr;
+        }
+
+        static constexpr _Ty* data() noexcept {
+            return nullptr;
+        }
+
+        static constexpr size_t size() noexcept {
+            return 0;
+        }
+
+        static constexpr bool empty() noexcept {
+            return true;
+        }
+    };
+
+    namespace views {
+        // VARIABLE TEMPLATE views::empty
+        template <class _Ty>
+        inline constexpr empty_view<_Ty> empty;
+    } // namespace views
+
+    // CLASS TEMPLATE ranges::single_view
+    // clang-format off
+    template <copy_constructible _Ty>
+        requires is_object_v<_Ty>
+    class single_view : public view_interface<single_view<_Ty>> {
+        // clang-format on
+    public:
+        single_view() = default;
+        constexpr explicit single_view(const _Ty& _Val_) noexcept(is_nothrow_copy_constructible_v<_Ty>) // strengthened
+            : _Val{in_place, _Val_} {}
+        constexpr explicit single_view(_Ty&& _Val_) noexcept(is_nothrow_move_constructible_v<_Ty>) // strengthened
+            : _Val{in_place, _STD move(_Val_)} {}
+
+        // clang-format off
+        template <class... _Types>
+            requires constructible_from<_Ty, _Types...>
+        constexpr explicit single_view(in_place_t, _Types&&... _Args) noexcept( // explicit per LWG-3428
+            is_nothrow_constructible_v<_Ty, _Types...>) // strengthened
+            // clang-format on
+            : _Val{in_place, _STD forward<_Types>(_Args)...} {}
+
+        _NODISCARD constexpr _Ty* begin() noexcept {
+            return data();
+        }
+        _NODISCARD constexpr const _Ty* begin() const noexcept {
+            return data();
+        }
+
+        _NODISCARD constexpr _Ty* end() noexcept {
+            return data() + 1;
+        }
+        _NODISCARD constexpr const _Ty* end() const noexcept {
+            return data() + 1;
+        }
+
+        _NODISCARD static constexpr size_t size() noexcept {
+            return 1;
+        }
+
+        _NODISCARD constexpr _Ty* data() noexcept {
+            return _STD addressof(*_Val);
+        }
+        _NODISCARD constexpr const _Ty* data() const noexcept {
+            return _STD addressof(*_Val);
+        }
+
+    private:
+        /* [[no_unique_address]] */ _Semiregular_box<_Ty> _Val{};

```

Comment:
I'm not sure any of these `[[no_unique_address]]`s add value given that `_Semiregular_box` will never be an empty type?

Reply:
The intent here is that `single_view<T>` is potentially zero-sized when `T` is empty. Note that `_Semiregular_box<T>` is potentially zero-sized when `T` is `semiregular` and potentially zero-sized by virtue of using `/* [[no_unique_address]] */` to declare `_Semiregular_box<T>`'s `T` member subobject.